importlib_metadata/__init__.py

```diff
@@ -1023,6 +1023,6 @@ def _top_level_declared(dist):
 def _top_level_inferred(dist):
     return {
         f.parts[0] if len(f.parts) > 1 else f.with_suffix('').name
-        for f in dist.files
+        for f in dist.files or []
```

Comment:
I see. I had simplified the summary of `always_iterable` to focus on the implementation for how it's used/recommended here. You're right that the implementation also addresses the concern of transforming a single item to an iterator over one item.

It solves the common pattern where a function accepts both a single item or an iterable of items but intends to iterate over them. For example consider [this change](https://github.com/jaraco/svg.charts/commit/eb7bcfe568e6cf10c977fda49e95f179e29bc7b7#diff-b5ee5757c7acd54889983e45eeb5701e1ada91dc4c236811c8a52c1cbe23c045L17-L33), where `always_iterable` transforms the branchy logic into a simple dict comprehension.

> That extra non-iterable condition is not expressible in mypy. 

That seems like a limitation of mypy. It seems to me that it should be possible to express any valid Python in mypy, especially something as simple as accepting an iterable or non-iterable. Is it possible the level of expressiveness you're trying to achieve is too great?  I notice that `always_iterable` does already have a [type specification](https://github.com/more-itertools/more-itertools/blob/31b8a0ed1c8a60d7c6b96bc6d2b7c64c646a1057/more_itertools/more.pyi#L285-L290). Would that spec suffice or is it too generic for strict mode?

Reply:
> It’s also a fragile condition to rely on

I think the same could be said for `or []`, which is also fragile based on the value of `ob.__bool__`. If an object were to return `False` for `__bool__` but not implement iterable, it also would fail. I've found that for all (most?) practical cases, iterability is stable for the uses of any particular function, despite being unreliable in the general case.

Reply:
> That seems like a limitation of mypy.

Yes. But there’s a good reason for this limitation, since it’s a consequence of the more important Liskov substitution principle: wherever an expression of type `T` is expected, it should be type-correct to provide a value whose type is a subclass of `T`.

Remember, every non-iterable type might be subclassed by an iterable type (this is not an uncommon corner case: *every* iterable type is a subclass of `object`). So if there were a hypothetical syntax that purported to guarantee that `always_iterable(obj): T` when `obj: T` and `T` is non-iterable, one could assign an instance of an iterable subclass `U` to `obj: T` and get back a result from `always_iterable(obj)` that breaks this guarantee.

A simpler way to say this is that there’s *no such thing* as a non-iterable type, provided that one recognizes a type to implicitly include all of its subclasses. Giving up that provision would have unpleasant ripple effects on every other aspect of the type system (I think you would basically need to augment each type annotation with a flag saying whether or not a subclass is allowed).

> I notice that always_iterable does already have a type specification.

`Any` is never a type specification; it just disables type checking for that value and every value derived from it. (See the [mypy documentation](https://mypy.readthedocs.io/en/stable/dynamic_typing.html) and [avoid the `Any` type](https://zulip.readthedocs.io/en/latest/testing/mypy.html#avoid-the-any-type).) Given that type checking is finding real bugs, we should be very hesitant to disable it for the sake of stylistic preferences.

There will always be some cases where appeasing the type checker rather than disabling it will mean writing code in a different way than you might have otherwise. No static type system can be powerful enough to capture all the ways a language as dynamic as Python can be used. I acknowledge that this is unfortunate. But I hope you’ll find as I have that the benefits of full type checking outweigh these inconveniences.

> I think the same could be said for `or []`, which is also fragile based on the value of `ob.__bool__`.

It’s conventional in Python, to the point of being enshrined in PEP 8, that testing a sequence as a boolean is an idiomatic check for emptiness. But let’s consider all the corner cases:

* A sequence that’s truthy but empty is fine since the result of `or []` is still empty.
* A sequence that’s falsy but nonempty would be treated as empty—given the PEP 8 idiom, if you write such a sequence, you should have expected that some users will treat it that way—but this doesn’t break type safety.
* There’s no value that’s `None` but truthy, since `NoneType` cannot be subclassed.

If that doesn’t satisfy you, here are some alternatives that type check:

```python
def _top_level_inferred(dist: Distribution) -> Set[str]:
    return {
        f.parts[0] if len(f.parts) > 1 else f.with_suffix('').name
        for f in ([] if dist.files is None else dist.files)
        if f.suffix == ".py"
    }
```

```python
def _top_level_inferred(dist: Distribution) -> Set[str]:
    if dist.files is None:
        return set()
    return {
        f.parts[0] if len(f.parts) > 1 else f.with_suffix('').name
        for f in dist.files
        if f.suffix == ".py"
    }
```

Perhaps the separate `if` statement is actually the clearest?